lib/nuca_backend/cats/cats.ex

```diff
@@ -3,26 +3,108 @@ defmodule NucaBackend.Cats do
   import Ecto.Query, warn: false
   alias NucaBackend.Repo
 
-  alias NucaBackend.Cats.Cat
+  alias NucaBackend.Cats.{Cat, Media.CatPicture}
 
   def list_cat do
     Repo.all(Cat) |> Repo.preload([:captured_by])
   end
 
-  def get_cat!(id), do: Repo.get!(Cat, id) |> Repo.preload([:captured_by])
+  def get_cat!(id), do: Repo.get!(Cat, id) |> Repo.preload([:captured_by, :media])
 
   def create_cat(attrs \\ %{}) do
-    %Cat{}
-    |> Cat.changeset(attrs)
-    |> Repo.insert()
-    |> OK.flat_map(fn (h) -> {:ok, Repo.preload(h, [:captured_by])} end)
+    result =
+      %Cat{}
+      |> Cat.changeset(attrs)
+      |> Repo.insert()
+
+    with {:result, {:ok, cat}} <- {:result, result},
+         {:preload, cat} <- {:preload, Repo.preload(cat, [:captured_by, :media])},
+         {:media, {:ok, media}} <- {:media, save_photos(cat, attrs["media"])},
+         {:match, cat} <-
+           {:match,
+            %{
+              cat
+              | media:
+                  media.new
+                  |> Enum.filter(fn {status, _media} -> status == :ok end)
+                  |> Enum.map(fn {:ok, media} -> media end)
+            }} do
+      {:ok, cat}
+    else
+      {:result, {:error, changeset}} -> {:error, changeset}
+      _ -> {:error, %{}}
+    end
   end
 
   def update_cat(%Cat{} = cat, attrs) do
-    cat
-    |> Cat.changeset(attrs)
-    |> Repo.update()
-    |> OK.flat_map(fn (h) -> {:ok, Repo.preload(h, [:captured_by])} end)
+    result = cat |> Cat.changeset(attrs) |> Repo.update()
+
+    with {:result, {:ok, cat}} <- {:result, result},
+         {:preload, cat} <- {:preload, Repo.preload(cat, [:captured_by, :media])},
+         {:media, {:ok, media}} <- {:media, save_photos(cat, attrs["media"])},
+         {:match, cat} <-
+           {:match,
+            %{
+              cat
+              | media:
+                  (media.new ++ cat.media)
+                  |> Enum.filter(fn
+                    {status, _media} -> status == :ok
+                    media -> media
+                  end)
+                  |> Enum.map(fn
+                    {:ok, media} -> media
+                    %CatPicture{} = picture -> picture
+                  end)
+            }} do
+      {:ok, cat}
+    else
+      {:result, {:error, changeset}} -> {:error, changeset}
+      _ -> {:error, %{}}
+    end
+  end
```

Comment:
I know this is not necessarily important, but this `with` statement seems to be very similar to the one in `create_cat` on lines 20-36. But considering that these functions will never really be reused, I don't think this is an issue overall.